core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1698,6 +1699,113 @@ RexNode simplifyAnd2ForUnknownAsFalse(List<RexNode> terms,
+    // the child node should be simplified
+    List<RexNode> operands = ((RexCall) node).getOperands();
+    List<RexNode> simplifiedChildren = operands.stream()
+        .map(this::eliminateCommonExprInCondition)
+        .collect(Collectors.toList());
+
+    switch (node.getKind()) {
+    case AND: {
+      // Split the children to get the disjunctive predicates.
+      List<List<RexNode>> disjunctions = simplifiedChildren.stream()
+          .map(RelOptUtil::disjunctions)
+          .collect(Collectors.toList());
+
+      // Find the common predict among the list by retainAll.
+      List<RexNode> common = new ArrayList<>(disjunctions.get(0));
+      disjunctions.forEach(common::retainAll);
+
+      if (common.isEmpty()) {
+        // no common, return the original but with child simplified
+        return and(simplifiedChildren);
+      }
+
+      // Get the predicates that without the common part.
+      disjunctions.forEach(p -> p.removeAll(common));
+
+      if (disjunctions.stream().anyMatch(List::isEmpty)) {
+        // (a || b) && (a || b || c || d) => (a || b)
+        return or(common);
+      }
+      // (a || b || c || d) && (a || b || e || f) =>
+      // (a || b) || ((c || d) && (e || f))
+      List<RexNode> splice = disjunctions.stream()
+          .map(this::or)
+          .collect(Collectors.toList());
+      return or(or(common), and(splice));
+    }
+    case OR: {
+      // Split the children to get the conjunctions predicates.
+      List<List<RexNode>> conjunctions = simplifiedChildren.stream()
+          .map(RelOptUtil::conjunctions)
+          .collect(Collectors.toList());
+
+      // Find the common predict among the list by retainAll.
+      List<RexNode> common = new ArrayList<>(conjunctions.get(0));
+      conjunctions.forEach(common::retainAll);
+
+      if (common.isEmpty()) {
+        // no common, return the original but with child simplified
+        return or(simplifiedChildren);
+      }
+
+      // Get the predicates that without the common part.
+      conjunctions.forEach(p -> p.removeAll(common));
+
+      if (conjunctions.stream().anyMatch(List::isEmpty)) {
+        // (a && b) || (a && b && c) => (a && b)
+        return and(common);
+      }
+      // (a && b && c && d) || (a && b && e && f) =>
+      // (a && b) && ((c && d) || (e && f))
+      List<RexNode> splice = conjunctions.stream()
+          .map(this::and)
+          .collect(Collectors.toList());
+      return and(and(common), or(splice));
+    }
+    default:
+      return node;
+    }
+  }
+
+  private RexNode and(RexNode... nodes) {
+    return and(ImmutableList.copyOf(nodes));
+  }
+
+  private RexNode and(List<RexNode> nodes) {
+    return RexUtil.composeConjunction(rexBuilder, nodes);
+  }
+
+  private RexNode or(RexNode... nodes) {
+    return or(ImmutableList.copyOf(nodes));
+  }
+
+  private RexNode or(List<RexNode> nodes) {
+    return RexUtil.composeDisjunction(rexBuilder, nodes);
+  }
```

Comment:
Is this the best place for the methods?

I'm inclined that the methods should be in `RexUtil` or something like that.

Reply:
Sure, I can move them to RexUtil, thanks for your advice.

Reply:
Emmmm there is a problem if I put these methods to RexUtil, we had to add RexBuilder to the method declare, pretty ugly, why I create these methods instead of using composeDisjunction/composeConjunction is for people to understand the cord more easily. 
Given that these methods are private, would it be OK to put them just in RexSimplify?